libraries/persistent_store/src/file.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! File-backed persistent flash storage for virtual authenticator.
+//!
+//! [`FileStorage`] implements the flash [`Storage`] interface but doesn't interface with an
+//! actual flash storage. Instead it uses a host-based file to persist the storage state.
+
+use crate::{Storage, StorageIndex, StorageResult};
+use alloc::borrow::Cow;
+use core::cell::RefCell;
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Seek, SeekFrom, Write};
+use std::path::Path;
+
+/// Simulates a flash storage using a host-based file.
+///
+/// This is usable for emulating authenticator hardware on VM hypervisor's host OS.
+pub struct FileStorage {
+    // Options of the storage.
+    options: FileOptions,
+
+    /// File for persisting contents of the storage.
+    ///
+    /// Reading data from File requires mutable reference, as seeking and reading data
+    /// changes file's current position.
+    ///
+    /// All operations on backing file internally always first seek to needed position,
+    /// so it's safe to borrow mutable reference to backing file for the time of operation.
+    file: RefCell<File>,
+}
+
+/// Options for file-backed storage.
+pub struct FileOptions {
+    /// Size of a word in bytes.
+    pub word_size: usize,
+
+    /// Size of a page in bytes.
+    pub page_size: usize,
+
+    /// Number of pages in storage.
+    pub num_pages: usize,
+}
+
+impl FileStorage {
+    pub fn new(path: &Path, options: FileOptions) -> StorageResult<FileStorage> {
+        let mut file_ref = RefCell::new(
+            OpenOptions::new()
+                .read(true)
+                .write(true)
+                .create(true)
+                .open(path)?,
+        );
+        let file = file_ref.get_mut();
+        let file_len = file.metadata()?.len();
+        let store_len: u64 = (options.page_size * options.num_pages) as u64;
+
+        if file_len == 0 {
+            file.seek(SeekFrom::Start(0))?;
+            let buf = vec![0xffu8; options.page_size];
```

Comment:
We have generally used `0xFF`. Is the `u8` annotation is required?

Other instances of this exist below.